src/multiplex/multiplex.rs

```diff
@@ -94,208 +165,255 @@ struct BodySender<T, B, E> {
+        while self.run {
+            // TODO: Only read frames if there is available space in the frame
+            // buffer
+            if let Async::Ready(frame) = try!(self.dispatch.transport().read()) {
+                try!(self.process_out_frame(frame));
+            } else {
+                break;
             }
-            Frame::MessageWithBody(id, out_message, body_sender) => {
-                trace!("   --> read out message; id={:?}", id);
+        }
+
+        Ok(())
+    }
+
+    /// Process outbound frame
+    fn process_out_frame(&mut self, frame: Frame<T::Out, T::BodyOut, T::Error>) -> io::Result<()> {
+        trace!("Multiplex::process_out_frame");
 
-                let body_sender = BodySender {
-                    deque: self.frame_buf.deque(),
-                    sender: body_sender.into(),
-                    is_ready: true,
-                };
+        match frame {
+            Frame::Message { id, message, body } => {
+                if body {
+                    let (tx, rx) = stream::channel();
+                    let tx = Sender::new(tx);
+                    let message = Message::WithBody(message, rx);
 
-                // Store the sender
-                self.out_bodies.insert(id, body_sender);
+                    try!(self.process_out_message(id, message, Some(tx)));
+                } else {
+                    let message = Message::WithoutBody(message);
 
-                // Process the actual message
-                try!(self.process_out_msg(id, Ok(out_message)));
+                    try!(self.process_out_message(id, message, None));
+                }
             }
-            Frame::Body(id, chunk) => {
+            Frame::Body { id, chunk } => {
                 trace!("   --> read out body chunk");
                 self.process_out_body_chunk(id, Ok(chunk));
             }
+            Frame::Error { id, error } => {
+                try!(self.process_out_err(id, error));
+            }
             Frame::Done => {
                 trace!("read Frame::Done");
-                assert!(self.in_bodies.len() == 0, "there are still in-bodies to process");
+                // TODO: Ensure all bodies have been completed
                 self.run = false;
             }
-            Frame::Error(id, err) => {
-                try!(self.process_out_err(id, err));
+        }
+
+        Ok(())
+    }
+
+    /// Process an outbound message
+    fn process_out_message(&mut self,
+                           id: RequestId,
+                           message: Message<T::Out, Body<T::BodyOut, T::Error>>,
+                           body: Option<Sender<T::BodyOut, T::Error>>)
+                           -> io::Result<()>
+    {
+        trace!("   --> process message; body={:?}", body.is_some());
+
+        match self.exchanges.entry(id) {
+            Entry::Occupied(mut e) => {
+                assert!(!e.get().responded, "invalid exchange state");
```

Comment:
Hm I'm actually not immediately sure how this is possible. Does this mean that the underlying multiplexer is reusing a request id for a new request, interrupting a previous one?

Or I guess put another way, what does it mean for a new message (a new request, right?) to come in while the request id already has an active exchange?